keps/0007-command-asserts.md

```diff
@@ -0,0 +1,103 @@
+---
+kep-number: 7
+short-desc: Assertions for CLI commands
+authors:
+  - "@nfnt"
+owners:
+  - "@nfnt"
+creation-date: 2020-10-28
+last-updated: 2020-10-28
+status: provisional
+---
+
+# Assertions for CLI commands
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [User Story 1](#user-story-1)
+    - [User Story 2](#user-story-2)
+  - [Implementation Details](#implementation-details)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+## Summary
+
+By asserting CLI commands, kuttl can support many additional test scenarios that examing the internal state of an operator.
+
+## Motivation
+
+Operators and custom resources can be used to deploy complex applications. For example, KUDO provides operators for Apache Cassandra and Apache Kafka. Deploying custom resources of these operators may not always result in changed Kubernetes resources but also in changes that aren't surfaced by Kubernetes resources but are internal to the operator. Asserting these changes only works through accessing service APIs or running commands and parsing their output. kuttl should provide tools to assert such output.
+
+### Goals
+
+- Assert that commands examining the internal state of an operator succeed or fail
+- Keep eventual consistency when asserting commands
```

Comment:
Can we specify what we mean with "eventual consistency" here? My understanding is that this means that all asserted things must match at some point in time - it's not enough that Assertion 1 matched at time T1 and fails at T2, and Assertion 2 fails at T1 and matches at T2. (In this case, both Assertions matched at *some* point in time, but never together.)

Is this what's meant by "eventual consistency" here?

Reply:
Yes, that's what's meant here. I'll clarify it in the KEP.